### ISSUE TITLE
Memory leak in 0.14.x

### ISSUE THREAD

**Issue by wubzz** [tgriesser/knex#2383]
# Environment

Knex version: 0.14.2
Database + version: Postgres 9.4
OS: Debian

# Bug

I've been experiencing quite the extensive memory leak since having updated my production app's knex version from 0.13 to 0.14. Leaving the app running from friday afternoon 'til monday morning resulted in ~500mb rss which never decreased after that.

After a lot of struggling and testing I managed to eventually get a pair of heapdumps that when compared still didn't make much sense but they did contain a lot of array and closure references to the underlying `generic-pool` module. This led me to believe the source was from knex, which also adds up to the changelog where [generic pool was upgraded](https://github.com/tgriesser/knex/pull/2208).

The pool options are as follows:
```javascript
connectionOptions.pool = {
	max: dbCfg.poolMax,
	min: 0,
	idleTimeoutMillis: 5000,
	evictionRunIntervalMillis: 1000,
	Promise,
};
```

Due to the nature of this leak being *'very slow but extensive'*, it is quite hard to reproduce as it takes a while for the memory to build up, and most likely relies on heavy traffic. For this reason I cannot guarantee that the leak is in `generic-pool` or if it is in `knex`, hence why this issue is not reported directly into the generic-pool repo.

After rolling back to v 0.13 of knex and leaving production app running for an equal amount of days this week, the app is still steady and fine at ~120mb, despite a lot more traffic during weekdays compared to weekends.

Has anyone else been experiencing something similar since upgrading to 0.14.2?

Unfortunately I'm pretty backed up with work until few weeks after new years, so I can't dig too much into this at this time. But I felt it neccessary to have this information available for others.

**Comment by wubzz** [comment]
Nothing really sticks out in our codebase [between 0.13 and 0.14](https://github.com/tgriesser/knex/compare/0.13.0...0.14.0) as potential memory leak (imo). Not sure how to proceed here, been trying to make a simple app to reproduce this but no luck so far.

**Comment by tgriesser** [comment]
I have a feeling this does have to do with the generic-pool bump to 3.x - there's a few issues that seem related including https://github.com/coopernurse/node-pool/issues/197 

I'm going to try to take some time this week to get caught up with the latest changes and see what we might want to do here. Generic-pool did some [pretty huge internal changes](https://github.com/coopernurse/node-pool/compare/v2.4.2...v3.1.7) (switching to Promises internally, swapping out the entire test suite, etc.), so I would place my bets on that being the cause. 

I'm considering just defaulting to utilizing the shipped implementation of the pool libraries for pg & mysql, and then using generic-pool 2.x for any other drivers that need it - that way we can offload any of these options to the underlying libs. 

This might be part of the first step in some bigger picture refactoring where we split the dialects out from the core so they're better tested & maintained. But I'll definitely report back with some thoughts & hopefully some PR's!

**Comment by wubzz** [comment]
@tgriesser +1000 for defaulting to client default when possible, that's something I would like to see happen.

I would happily take the time to implement this for `pg` as a starting point if you're on board.

**Comment by wubzz** [comment]
Seems like a pretty straight-forward almost 'drop-in-replacement' to use built-in `pg` pooling rather than `generic-pool`. @tgriesser Would you be OK with merging this change now instead of waiting for the refactoring-overhaul?

**Comment by falnos24865** [comment]
I can confirm there is something wrong here. I ran a small test that did nothing but run a SELECT using knex repeatedly and monitored the heap usage after each query. Knex.js was burning anywhere from 700k to 2MB of heap per query, which crashed after a few hundred queries when it hit the node heap limit at around 1.5gb.

**Comment by elhigu** [comment]
@falnos24865 please share your test app

**Comment by igor-savin-ht** [comment]
@falnos24865 Can you run same test against tgriesser/knex#master npm version? I wonder if latest updates to generic-pool resolved the leak.

**Comment by elhigu** [comment]
I'll try to write test for this one.

**Comment by elhigu** [comment]
@falnos24865 @wubzz which DB are you using? I have not been able to reproduce any memory leak so far. That @falnos24865 reported functionality couldn't be reproduced. But I did find a problem where dead connections are not returned to pool or something like that but it did seem pretty dangerous and made knex to hang and just wait. Need to look more to it.

**Comment by elhigu** [comment]
I was able to reproduce memory leaking on postgres with master that doesn't happen in 0.13 with this code:

```js
const Bluebird = require('bluebird');

const pg = require('../lib')({
  client: 'pg',
  connection: 'postgres:///knex_test',
  pool: { max: 90 }
});

/**
const mysql = require('../lib')({
  dialect: 'mysql2',
  connection: {
    database: "knex_test",
    user: "root",
    charset: 'utf8'
  }
});
 */

async function main() {
  while(true) {
    try {
      await Promise.all([
        ...Array(80).fill(pg('accounts').timeout(100)),
        pg.raw(`SELECT pg_terminate_backend(pg_stat_activity.pid) FROM pg_stat_activity WHERE pg_stat_activity.datname = 'knex_test' AND pid <> pg_backend_pid()`),
      ]);
    } catch (err) {
      console.log(err);
      // pass on error and try again...
    }
  }
}

main();
```

It is running lots of queries and killing the connections at the same time from server side.

**Comment by elhigu** [comment]
And running the same code with https://github.com/tgriesser/knex/pull/2450 branch doesn't seem to leak.

**Comment by igor-savin-ht** [comment]
@elhigu Great news! Any chance for an early release after #2450 lands, e. g. in RC form?

**Comment by elhigu** [comment]
I just release straight 0.14.3 later on this week.

**Comment by elhigu** [comment]
@knownasilya if you are able to create repro of that, please open new issue. If you are not able to isolate the issue to knex reason for it might be anywhere.

**Comment by calvinmetcalf** [comment]
this is actually a node-postgres bug it looks like brianc/node-postgres#1417
